File: 5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
#include<string>
#include<vector>
//#include<iostream>
using namespace std;
class Solution {
public:
    string longestPalindrome(string s) {
        vector<vector<bool>> isPalindromic(s.size(),vector<bool>(s.size(),true));

        int max_len = 1;
        int start_pos = 0;
        for (int i = 1; i < s.size(); i++)
        {
            for (int j = 0; j < s.size()-i; j++)
            {
                //[j][j+1]
                if(isPalindromic[j+1][j+i-1] && (s[j] == s[j+i])){
                    // isPalindromic[j][j+i] = true;
                    if(i+1 > max_len){
                        max_len = i+1;
                        start_pos = j;
                    }
                }
                else
                    isPalindromic[j][j+i] = false;
            }
        }

        return s.substr(start_pos,max_len);
    }

    // void print(vector<vector<bool>> &isPalindromic){
    //     for (size_t i = 0; i < isPalindromic.size(); i++)
    //     {
    //         for (size_t j = 0; j < isPalindromic.size(); j++)
    //         {
    //             cout<<isPalindromic[i][j]<<'|';
    //             /* code */
    //         }
    //         cout<<endl;
    //         /* code */
    //     }
    //     cout<<endl<<"============="<<endl;
    // }

};
// @lc code=end

File: 5.longest-palindromic-substring.go
/*
 * @lc app=leetcode id=5 lang=golang
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
var memo map[int]int

var _s string

func longestPalindrome(s string) string {
	_s = s

	memo = make(map[int]int, 0)
	mlen := 0
	res := ""
	for i := 0; i < len(s); i++ {
		len := dp(i)
		if len > mlen {
			mlen = len
			res = s[i-len+1 : i+1]
		}
	}

	return res
}

func dp(j int) int {
	if j == 0 {
		return 1
	}

	if v, ok := memo[j]; ok {
		return v
	}

	mlen := 1
	i := j - dp(j-1) - 1
	for k := max(0, i); k < j; k++ {
		if _s[k] == _s[j] && isPalindrome(k, j) {
			return j - k + 1
		}
	}

	memo[j] = mlen
	return mlen
}

func isPalindrome(i, j int) bool {
	for i < j {
		if _s[i] != _s[j] {
			return false
		}
		i++
		j--
	}

	return true
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

// @lc code=end
